src/main/cpp/src/regex_rewrite_utils.cu

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/find.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/atomic>
+#include <thrust/binary_search.h>
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+namespace spark_rapids_jni {
+
+/**
+ * @brief Utility to return a bool column indicating the presence of
+ * a given prefix string followed by at leasat range_len characters
+ * with code points in the range [start, end] in each string of the input column.
+ *
+ * @tparam BoolFunction Return bool value given two strings.
+ *
+ * @param strings Column of strings to check for prefix.
+ * @param prefix UTF-8 encoded string to check in strings column.
+ * @param range_len Minimum number of characters to check after the prefix.
+ * @param start Minimum code point value to check for in the range.
+ * @param end Maximum code point value to check for in the range.
+ * @param pfn Returns bool value if prefix is found in the given string.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory.
+ * @return New BOOL column.
+ */
+template <typename BoolFunction>
+std::unique_ptr<cudf::column> literal_range_pattern_fn(cudf::strings_column_view const& strings,
```

Comment:
Functions/classes used only within the current file need to be wrapped in anonymous namespace to avoid symbol conflict.